File: rust/src/metaparser.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

use lidy__yaml::{LineCol, Yaml, YamlData};

use crate::error::{AnyBoxedError, JoinError, SimpleError};
use crate::file::File;
use crate::lidy::{make_rule_set, Builder, Parser, ParserData};
use crate::result::{Data, LidyResult, Position};
use crate::rule::{apply_predefined_rule, Rule};
use crate::yamlfile::YamlFile;

fn check_merged_node<TV, TB>(
    name: &str,
    last_position: &Position,
    origin_position: &Position,
    subparser: &Parser<TV, TB>,
) -> Option<AnyBoxedError>
where
    TV: Clone,
    TB: Builder<TV>,
{
    let rule = subparser.rule_set.get(name);
    if rule.is_none() {
        return Some(Box::new(SimpleError::from_check_result(
            "_merge",
            &format!(
                "unknown rule '{name}' encountered at {} following rules from a _merge keyword",
                LineCol::from(last_position)
            ),
            (origin_position).clone().into(),
        )));
    }
    // check_error, to be returned only if the node is not a map checker
    let check_error = SimpleError::from_check(
        "_merge",
        "reference leads to a non-map-checker node",
        &rule.as_ref()?.node,
    );

    let rule_node = &rule.unwrap().node;
    let last_pos = Position::from_line_col_beginning_only(
        origin_position.filename.clone(),
        rule_node.line_col,
    );

    match &rule_node.data {
        YamlData::String(name) => check_merged_node(name, &last_pos, origin_position, subparser),
        YamlData::Mapping(map) => {
            let is_map_checker = map.iter().any(|(key, _)| {
                if let YamlData::String(key_str) = &key.data {
                    matches!(
                        key_str.as_str(),
                        "_map" | "_mapFacultative" | "_mapOf" | "_merge"
                    )
                } else {
                    false
                }
            });
            if !is_map_checker {
                Some(Box::new(check_error))
            } else {
                None
            }
        }
        _ => Some(Box::new(check_error)),
    }
}

pub fn make_meta_parser_for<TV, TB>(
    subparser: &mut Parser<TV, TB>,
) -> Result<Parser<TV, TB>, AnyBoxedError>
where
    TV: Clone + 'static,
    TB: Builder<TV>,
{
    let meta_schema_file = File::read_local_file("../../lidy.schema.yaml")?;
    let mut meta_schema = YamlFile::new(Rc::new(meta_schema_file));
    meta_schema.unmarshal()?;

    let subparser_ref = Rc::new(RefCell::new(subparser));

    let rule_reference_builder: Builder<TV> = {
        // let subparser_ref = Rc::clone(&subparser_ref);
        let builder_fn = Box::new(
            |input: &LidyResult<TV>| -> Result<Data<TV>, AnyBoxedError> {
                let identifier = match &input.data {
                    Data::String(s) => s.to_string(),
                    _ => return Ok(input.data.clone()),
                };

                if identifier == "expression" {
                    println!(
                        "identifier === 'expression', stack: {:?}",
                        std::backtrace::Backtrace::capture()
                    );
                }

                let mut subparser = subparser_ref.borrow_mut();
                if let Some(rule) = subparser.rule_set.get_mut(&*identifier) {
                    rule.is_used = true;
                } else {
                    let rule_exists = match apply_predefined_rule(
                        &ParserData::<(), ()>::default(),
                        &identifier,
                        &Yaml::default(),
                        true,
                    )?.data {
                        Data::Boolean(exists) => exists,
                        _ => panic!("never, apply_predefined_rule must return a boolean when only_check_if_rule_exists is passed")
                    };
                    if !rule_exists {
                        let rule_listing = subparser
                            .rule_set
                            .keys()
                            .map(|k| k.to_string())
                            .collect::<Vec<_>>()
                            .join(", ");
                        return Err(Box::new(SimpleError::from_check_result(
                            &identifier,
                            &format!(
                                "encountered unknown rule identifier '{}'. Known rules are: [{}]",
                                identifier, rule_listing
                            ),
                            LineCol {
                                line: input.position.line,
                                column: input.position.column,
                            },
                        )));
                    }
                }
                Ok(input.data.clone())
            },
        )
            as Box<dyn FnMut(&LidyResult<TV>) -> Result<Data<TV>, Box<(dyn std::error::Error)>>>;
        Rc::new(RefCell::new(builder_fn))
    };

    let map_checker_builder: Builder<TV> = {
        let subparser_ref = Rc::clone(&subparser_ref);
        let builder_fn = Box::new(
            |input: &LidyResult<TV>| -> Result<Data<TV>, AnyBoxedError> {
                if let Data::MapData(map_data) = &input.data {
                    if let Some(merge) = map_data.map.get("_merge") {
                        let mut join_error = JoinError::default();
                        if let Data::ListData(list_data) = &merge.data {
                            let subparser = subparser_ref.borrow();
                            for result in &list_data.list_of {
                                match &result.data {
                                    Data::CustomData(_) => continue,
                                    Data::String(s) => {
                                        if let Some(err) = check_merged_node(
                                            s,
                                            &result.position,
                                            &result.position,
                                            &subparser,
                                        ) {
                                            join_error.add(err);
                                        }
                                    }
                                    _ => continue,
                                }
                            }
                        }
                        join_error.into_result()?;
                    }
                }
                Ok(input.data.clone())
            },
        )
            as Box<dyn FnMut(&LidyResult<TV>) -> Result<Data<TV>, Box<(dyn std::error::Error)>>>;
        Rc::new(RefCell::new(builder_fn))
    };

    let sized_checker_keyword_set_builder: Builder<TV> = {
        let builder_fn = Box::new(
            |input: &LidyResult<TV>| -> Result<Data<TV>, AnyBoxedError> {
                if let Data::MapData(map_data) = &input.data {
                    for keyword in &["_min", "_max", "_nb"] {
                        if let Some(value) = map_data.map.get(*keyword) {
                            if let Data::Float(n) = &value.data {
                                if *n < 0.0 {
                                    return Err(Box::new(SimpleError::from_check_result(
                                        keyword,
                                        "cannot be negative",
                                        LineCol {
                                            line: input.position.line,
                                            column: input.position.column,
                                        },
                                    )));
                                }
                            }
                        }
                    }
                    let min = map_data.map.get("_min").and_then(|v| match v.data {
                        Data::Float(n) => Some(n),
                        _ => None,
                    });
                    let max = map_data.map.get("_max").and_then(|v| match v.data {
                        Data::Float(n) => Some(n),
                        _ => None,
                    });
                    let nb = map_data.map.get("_nb").and_then(|v| match v.data {
                        Data::Float(n) => Some(n),
                        _ => None,
                    });

                    if nb.is_some() {
                        let mut min_or_max = "";
                        if min.is_some() {
                            min_or_max = "min";
                        } else if max.is_some() {
                            min_or_max = "max";
                        }

                        return Err(Box::new(SimpleError::from_check_result(
                            "_nb",
                            &format!(
                                "it makes no sense to use the `_nb` and `_{min_or_max}` together"
                            ),
                            LineCol {
                                line: input.position.line,
                                column: input.position.column,
                            },
                        )));
                    }

                    if let (Some(min), Some(max)) = (min, max) {
                        if min > max {
                            return Err(Box::new(SimpleError::from_check_result(
                                "_min",
                                "`_max` cannot be lower than `_min`",
                                LineCol {
                                    line: input.position.line,
                                    column: input.position.column,
                                },
                            )));
                        }
                    }
                }
                Ok(input.data.clone())
            },
        )
            as Box<dyn FnMut(&LidyResult<TV>) -> Result<Data<TV>, Box<(dyn std::error::Error)>>>;
        Rc::new(RefCell::new(builder_fn))
    };

    let meta_builder_map: HashMap<Box<str>, Builder<TV>> = HashMap::from([
        (Box::from("ruleReference"), rule_reference_builder),
        (Box::from("mapChecker"), map_checker_builder),
        (
            Box::from("sizedCheckerKeywordSet"),
            sized_checker_keyword_set_builder,
        ),
    ]);

    let rule_set = make_rule_set(&meta_schema, meta_builder_map)?;
    Ok(Parser { rule_set })
}

pub fn check_rule_set(rule_set: &mut HashMap<Box<str>, Rule>) -> Result<(), AnyBoxedError> {
    let mut join_error = JoinError::default();

    // Check main rule exists and mark it as used
    if let Some(main_rule) = rule_set.get_mut("main") {
        main_rule.is_used = true;
    } else {
        join_error.add(Box::new(SimpleError::from_str(
            "could not find the 'main' rule",
        )));
    }

    // Check for unused rules
    for (name, rule) in rule_set.iter() {
        if !rule.is_used {
            join_error.add(Box::new(SimpleError::from_str(&format!(
                "rule '{}' is defined but never used",
                name
            ))));
        }
    }

    // Check for direct rule references
    for (name, rule) in rule_set.iter() {
        if let Some(err) = check_direct_rule_reference(rule_set, &rule.node, &[&**name]) {
            join_error.add(err);
        }
    }

    join_error.into_result()
}

fn check_direct_rule_reference(
    rule_set: &HashMap<Box<str>, Rule>,
    rule_node: &Yaml,
    rule_name_array: &[&str],
) -> Option<AnyBoxedError> {
    match &rule_node.data {
        YamlData::String(s) => {
            // Check for self-reference
            if rule_name_array.contains(&s.as_str()) {
                return Some(Box::new(SimpleError::from_str(&format!(
                    "rule '{}' references itself",
                    s
                ))));
            }

            // Check target rule
            if let Some(target_rule) = rule_set.get(&**s) {
                check_direct_rule_reference(
                    rule_set,
                    &target_rule.node,
                    &rule_name_array
                        .iter()
                        .chain(std::iter::once(&s.as_str()))
                        .copied()
                        .collect::<Vec<_>>(),
                )
            } else {
                None // Predefined rule
            }
        }
        YamlData::Mapping(map) => {
            let mut join_error = JoinError::default();

            // Check for _oneOf or _merge nodes
            for (key, value) in map {
                if let YamlData::String(key_str) = &key.data {
                    if key_str == "_oneOf" || key_str == "_merge" {
                        if let YamlData::List(seq) = &value.data {
                            for node in seq {
                                if let Some(err) =
                                    check_direct_rule_reference(rule_set, node, rule_name_array)
                                {
                                    join_error.add(err);
                                }
                            }
                        }
                    }
                }
            }

            match join_error.into_result() {
                Ok(_) => None,
                Err(e) => Some(e),
            }
        }
        _ => Some(Box::new(SimpleError::from_str(
            "rule node should be either a scalar or a mapping",
        ))),
    }
}

File: js/src/metaparser.ts
import * as yaml from "yaml"

import { CheckError, CheckResultError, JoinError } from "./error"
import { readLocalFile } from "./file"
import { Builder, makeParserFromRuleSet, makeRuleSet, Parser } from "./lidy"
import { ListData, MapData, Position, Result } from "./result"
import { applyPredefinedRule, Rule } from "./rule"
import * as syaml from "./syaml"
import { deserializedYamlFile, YamlFile } from "./yamlfile"

export function makeMetaParserFor(subparser: Parser): Parser {
  const metaSchema: YamlFile = {
    file: readLocalFile("../../lidy.schema.yaml"),
    yaml: new yaml.YAMLMap(),
    doneParsing: false,
    lineCounter: new yaml.LineCounter(),
  }
  const error = deserializedYamlFile(metaSchema)
  if (error) {
    throw error
  }

  const checkMergedNode = (
    name: string,
    namePosition: Position,
    originPosition: Position,
  ): Error | undefined => {
    const rule = subparser.ruleSet[name]
    // checkError, to be returned only if the node is not a map checker
    const checkError = new CheckError(
      "_merge",
      `reference leads to a non-map-checker node`,
      metaParserData as any,
      rule.node,
    )

    if (!rule) {
      return new CheckResultError(
        "_merge",
        `unknown rule '${name}' at ${namePosition} encountered followings rules from a _merge keyword`,
        originPosition,
      )
    } else if (yaml.isScalar(rule.node)) {
      if (!syaml.isString(rule.node)) {
        return checkError
      }
      const { line, col: column } = metaParserData.lineCounter.linePos(
        rule.node.range?.[0] || 0,
      )
      const { line: lineEnd, col: columnEnd } =
        metaParserData.lineCounter.linePos(rule.node.range?.[1] || 0)
      const newNamePosition = {
        filename: originPosition.filename,
        line,
        column,
        lineEnd,
        columnEnd,
      }
      return checkMergedNode(rule.node.value, newNamePosition, originPosition)
    } else if (yaml.isMap<any, yaml.Node>(rule.node)) {
      const isMapChecker = rule.node.items.some(({ key }) => {
        if (
          key.value === "_map" ||
          key.value === "_mapFacultative" ||
          key.value === "_mapOf" ||
          key.value === "_merge"
        ) {
          return true
        }
      })
      if (!isMapChecker) {
        return checkError
      }
    }
  }

  const metaBuilderMap: Record<string, Builder> = {
    ruleReference: (input: Result<any>) => {
      const identifier = input.data
      if (identifier === "expression") {
        console.log("identifier === 'expression', stack:", new Error().stack)
      }
      const rule = subparser.ruleSet[identifier]
      if (!rule) {
        const $$onlyCheckIfRuleExistsRef = { value: false }
        applyPredefinedRule(
          {} as any,
          identifier,
          {} as any,
          $$onlyCheckIfRuleExistsRef,
        )

        // If the rule does not exist, we throw an error
        if (!$$onlyCheckIfRuleExistsRef.value) {
          const ruleListing = Object.keys(subparser.ruleSet).join(", ")
          throw new CheckResultError(
            identifier,
            `encountered unknown rule identifier '${identifier}'. Known rules are: [${ruleListing}]`,
            input.position,
          )
        }
        // Otherwise, nothing special to do, we can just return the input
      } else {
        rule.isUsed = true
      }
      return {
        data: input.data,
        isLidyData: true,
      }
    },
    mapChecker: (input: Result<any>) => {
      const mapData = input.data as MapData<any, string | ListData<any>>
      const _merge = mapData.map["_merge"]
      const joinError = new JoinError()
      if (_merge) {
        const mergedNodeSlice = (_merge.data as ListData<any>).listOf
        mergedNodeSlice.forEach((result) => {
          if (result.data.isLidyData && result.data._map) {
            // the merged node is a _map* checker, it's okay
            return // continue
          } else if (typeof result.data === "string") {
            // the merged node is an identifier, the rule it refers to.
            joinError.add(
              checkMergedNode(result.data, result.position, result.position),
            )
          }
        })
      }
      joinError.throw()
      return {
        data: input.data,
        isLidyData: true,
      }
    },
    sizedCheckerKeywordSet: (input: Result<MapData<any, any>>) => {
      const mapData = input.data
      ;["_min", "_max", "_nb"].forEach((keyword) => {
        const value = mapData.map[keyword]
        if (value && value.data < 0) {
          throw new CheckResultError(
            keyword,
            "cannot be negative",
            input.position,
          )
        }
      })
      const _min = mapData.map["_min"]
      const _max = mapData.map["_max"]
      const _nb = mapData.map["_nb"]
      if (_nb && _min) {
        throw new CheckResultError(
          "_nb",
          "it makes no sense to use the `_nb` and `_min` together",
          input.position,
        )
      }
      if (_nb && _max) {
        throw new CheckResultError(
          "_nb",
          "it makes no sense to use the `_nb` and `_max` together",
          input.position,
        )
      }
      if (_min && _max && _min.data > _max.data) {
        throw new CheckResultError(
          "_min",
          "`_max` cannot be lower than `_min`",
          input.position,
        )
      }
      return {
        data: input.data,
        isLidyData: true,
      }
    },
  }

  const metaRuleSet = makeRuleSet(metaSchema, metaBuilderMap)

  const metaParser = makeParserFromRuleSet(metaRuleSet)

  const metaParserData = {
    contentFileName: metaSchema.file.name,
    lineCounter: metaSchema.lineCounter,
    parser: metaParser,
    ruleTrace: [],
  }

  return metaParser
}

function checkDirectRuleReference(
  ruleSet: Record<string, Rule>,
  ruleNode: yaml.Node,
  ruleNameArray: string[],
): Error | undefined {
  if (yaml.isScalar(ruleNode)) {
    for (const ruleName of ruleNameArray) {
      if (ruleNode.value === ruleName) {
        return new Error(`rule '${ruleName}' references itself`)
      }
    }
    const targetRule = ruleSet[ruleNode.value as string]
    if (targetRule) {
      return checkDirectRuleReference(ruleSet, targetRule.node, [
        ...ruleNameArray,
        ruleNode.value as string,
      ])
    } else {
      // The rule is a predefined rule
      return undefined
    }
  } else if (!yaml.isMap<string, yaml.YAMLSeq<yaml.Node>>(ruleNode)) {
    throw new Error("rule node should be either a scalar or a mapping")
  }
  const directChildNodePair = ruleNode.items.find(({ key }) => {
    return (
      yaml.isScalar(key) && (key.value === "_oneOf" || key.value === "_merge")
    )
  })
  const joinError = new JoinError()
  if (directChildNodePair) {
    directChildNodePair.value!.items.forEach((node) => {
      joinError.add(checkDirectRuleReference(ruleSet, node, ruleNameArray))
    })
  }
  return joinError
}

export function checkRuleSet(ruleSet: Record<string, Rule>) {
  const joinError = new JoinError()
  const mainRule = ruleSet["main"]
  if (!mainRule) {
    joinError.add(new Error("could not find the 'main' rule"))
  } else {
    mainRule.isUsed = true
  }
  Object.entries(ruleSet).forEach(([name, rule]) => {
    if (!rule.isUsed) {
      joinError.add(new Error(`rule '${name}' is defined but never used`))
    }
  })
  Object.entries(ruleSet).forEach(([name, rule]) => {
    joinError.add(checkDirectRuleReference(ruleSet, rule.node, [name]))
  })
  joinError.throw()
}
